### ISSUE TITLE
WGSL: u32 is not cast to f32 when building Vec2 from u32

### ISSUE THREAD

**Issue by rishflab** [gfx-rs/wgpu#5020]
**Description**

```
let a = vec2<f32>(10u);
```

This code passes the validation check but `a` ends up being a `vec2<u32>` instead of a `vec2<f32>` as expected.

I know that it is a `vec2<u32>` because when I explicitly state the desired type in the binding like so
```
   let a: vec2<f32> = vec2<f32>(10u);
``` 
I get the following error:

```
Shader 'Accumulate' parsing error: the type of `a` is expected to be `vec2<f32>`, but got `vec2<u32>`
    ┌─ wgsl:145:13
    │
145 │         let a: vec2<f32> = vec2<f32>(10u);
    │             ^^^^^^^^^^^ definition of `a`


    the type of `a` is expected to be `vec2<f32>`, but got `vec2<u32>`

```

**Expected Behaviour**

```
let a = vec2<f32>(10u);
```

The constructor should cast the u32 to a f32 and produce a `vec2<f32>` 
OR 
this line should fail validation if `u32` is not a valid argument to the `vec2<f32>` constructor.

**Platform**
Windows 10, wgpu 0.17

**Comment by ErichDonGubler** [comment]
I believe that this specific case should be resolved with recent work to make constructor expressions constant-evaluatable? I'm not super familiar with that work, so I might be wrong.

**Comment by teoxoy** [comment]
@rishflab could you try latest trunk? According to the spec the examples provided should all lead to validation errors.

As Erich pointed out, we recently touched the code around constructors so it could be that all the examples are now correctly rejected.

**Comment by rishflab** [comment]
The bug is fixed on trunk/4400a5847080d1164bdca93a90622414963ed9f3
The bug is present on 0.18.2 and 0.17.0

**Comment by cwfitzgerald** [comment]
Nice, then this can be closed - 0.19 will be out next week with the fix. I don't believe any of these fixes are easy to backport, unfortunately.
